lightwood/encoders/text/pretrained.py

```diff
@@ -0,0 +1,362 @@
+                """
+                Freeze the base transformer model and train
+                a linear layer on top
+                """
+                # Freeze all the transformer parameters
+                for param in self._model.base_model.parameters():
+                    param.requires_grad = False
+
+                optimizer_grouped_parameters = self._model.parameters()
+
+            else:
+                print("\tFine-tuning model")
+                """
+                Fine-tuning parameters with weight decay
+                """
+                # Fine-tuning weight-decay (https://huggingface.co/transformers/training.html)
+                no_decay = [
+                    "bias",
+                    "LayerNorm.weight",
+                ]  # no decay on the classifier terms.
+                optimizer_grouped_parameters = [
+                    {
+                        "params": [
+                            p
+                            for n, p in self._model.named_parameters()
+                            if not any(nd in n for nd in no_decay)
+                        ],
+                        "weight_decay": 0.01,
+                    },
+                    {
+                        "params": [
+                            p
+                            for n, p in self._model.named_parameters()
+                            if any(nd in n for nd in no_decay)
+                        ],
+                        "weight_decay": 0.0,
+                    },
+                ]
+
+            optimizer = AdamW(optimizer_grouped_parameters, lr=1e-5)
+            scheduler = get_linear_schedule_with_warmup(
+                optimizer,
+                num_warmup_steps=0,  # default value for GLUE
+                num_training_steps=len(dataset) * self._epochs,
+            )
+
+            # Train model; declare optimizer earlier if desired.
+            self._train_model(
+                dataset, optim=optimizer, scheduler=scheduler, n_epochs=self._epochs
+            )
+
+        else:
+            print("Embeddings Generator only")
+
+            self.model_type = "embeddings_generator"
+            self._model = self._embeddings_model_class.from_pretrained(
+                self._pretrained_model_name
+            ).to(self.device)
+
+        self._prepared = True
+
+    def _train_model(self, dataset, optim=None, scheduler=None, n_epochs=4):
+        """
+        Given a model, train for n_epochs.
+
+        model - torch.nn model;
+        dataset - torch.DataLoader; dataset to train
+        device - torch.device; cuda/cpu
+        log - lightwood.logger.log; print output
+        optim - transformers.optimization.AdamW; optimizer
+        n_epochs - number of epochs to train
+
+        """
+        self._model.train()
+
+        if optim is None:
+            print("Setting all model params to AdamW")
+            optim = AdamW(self._model.parameters(), lr=5e-5)
+
+        for epoch in range(n_epochs):
```

Comment:
It's not strictly needed, but we could try to use `LightwoodAutocast` (unless HuggingFace automatically handles it internally). You can find an example [here](https://github.com/mindsdb/lightwood/blob/a8e5f7678a6aef0ea233a35f30a58107a0913484/lightwood/encoders/time_series/rnn.py#L174).